File: Practica 11/Java/Main.java
package Java;

interface Figure {
    String getColor();
    double perimeter();
    double area();
}

class Triangle implements Figure {
    private String color;
    private double a, b, c;

    public Triangle(String color, double a, double b, double c) {
        this.color = color;
        this.a = a;
        this.b = b;
        this.c = c;
    }

    @Override
    public String getColor() {
        return color;
    }

    @Override
    public double perimeter() {
        return a + b + c;
    }

    @Override
    public double area() {
        double s = perimeter() / 2;
        return Math.sqrt(s * (s - a) * (s - b) * (s - c));
    }
}

class Circle implements Figure {
    private String color;
    private double radius;

    public Circle(String color, double radius) {
        this.color = color;
        this.radius = radius;
    }

    @Override
    public String getColor() {
        return color;
    }

    @Override
    public double perimeter() {
        return 2 * Math.PI * radius;
    }

    @Override
    public double area() {
        return Math.PI * radius * radius;
    }
}

class Rectangle implements Figure {
    private String color;
    private double width, height;

    public Rectangle(String color, double width, double height) {
        this.color = color;
        this.width = width;
        this.height = height;
    }

    @Override
    public String getColor() {
        return color;
    }

    @Override
    public double perimeter() {
        return 2 * (width + height);
    }

    @Override
    public double area() {
        return width * height;
    }
}

class Hexagon implements Figure {
    private String color;
    private double sideLength;

    public Hexagon(String color, double sideLength) {
        this.color = color;
        this.sideLength = sideLength;
    }

    @Override
    public String getColor() {
        return color;
    }

    @Override
    public double perimeter() {
        return 6 * sideLength;
    }

    @Override
    public double area() {
        return (3 * Math.sqrt(3) * sideLength * sideLength) / 2;
    }
}

public class Main {
    public static void main(String[] args) {
        Figure[] shapes = {
            new Triangle("Red", 3, 4, 5),
            new Circle("Blue", 2.5),
            new Rectangle("Green", 4, 6),
            new Hexagon("Yellow", 2)
        };

        for (Figure shape : shapes) {
            System.out.println("Shape: " + shape.getClass().getSimpleName() +
                               ", Color: " + shape.getColor() +
                               ", Perimeter: " + shape.perimeter() +
                               ", Area: " + shape.area());
        }
    }
}

File: Practica 11/Python/Main.py
from abc import ABC, abstractmethod
from math import sqrt, pi

class Figure(ABC):
    @abstractmethod
    def get_color(self):
        pass

    @abstractmethod
    def perimeter(self):
        pass

    @abstractmethod
    def area(self):
        pass

class Triangle(Figure):
    def __init__(self, color, a, b, c):
        self.color = color
        self.a = a
        self.b = b
        self.c = c

    def get_color(self):
        return self.color

    def perimeter(self):
        return self.a + self.b + self.c

    def area(self):
        s = self.perimeter() / 2
        return sqrt(s * (s - self.a) * (s - self.b) * (s - self.c))

class Circle(Figure):
    def __init__(self, color, radius):
        self.color = color
        self.radius = radius

    def get_color(self):
        return self.color

    def perimeter(self):
        return 2 * pi * self.radius

    def area(self):
        return pi * self.radius ** 2

class Rectangle(Figure):
    def __init__(self, color, width, height):
        self.color = color
        self.width = width
        self.height = height

    def get_color(self):
        return self.color

    def perimeter(self):
        return 2 * (self.width + self.height)

    def area(self):
        return self.width * self.height

class Hexagon(Figure):
    def __init__(self, color, side_length):
        self.color = color
        self.side_length = side_length

    def get_color(self):
        return self.color

    def perimeter(self):
        return 6 * self.side_length

    def area(self):
        return (3 * sqrt(3) * self.side_length ** 2) / 2

# Demonstration of polymorphism
shapes = [
    Triangle("Red", 3, 4, 5),
    Circle("Blue", 2.5),
    Rectangle("Green", 4, 6),
    Hexagon("Yellow", 2)
]

for shape in shapes:
    print(f"Shape: {shape.__class__.__name__}, Color: {shape.get_color()}, Perimeter: {shape.perimeter()}, Area: {shape.area()}")
